File: src/nowifi/report.py
"""Reporting: terminal, markdown, and JSON output."""

from __future__ import annotations

import json
from datetime import datetime, timezone

from rich.console import Console
from rich.panel import Panel
from rich.table import Table

from .bypass import BypassResult, Severity
from .detect import PortalInfo
from .probe import ProbeResults


def _severity_style(severity: Severity) -> str:
    return {
        Severity.CRITICAL: "bold red",
        Severity.HIGH: "red",
        Severity.MEDIUM: "yellow",
        Severity.LOW: "cyan",
        Severity.INFO: "dim",
    }.get(severity, "dim")


def _bool_icon(val: bool) -> str:
    return "[green]OPEN[/green]" if val else "[red]CLOSED[/red]"


def _severity_icon(severity: Severity) -> str:
    icons = {
        Severity.CRITICAL: "[bold red]CRIT[/bold red]",
        Severity.HIGH: "[red]HIGH[/red]",
        Severity.MEDIUM: "[yellow]MED[/yellow]",
        Severity.LOW: "[cyan]LOW[/cyan]",
        Severity.INFO: "[dim]INFO[/dim]",
    }
    return icons.get(severity, "[dim]?[/dim]")


def print_terminal_report(
    portal_info: PortalInfo,
    probe_results: ProbeResults,
    bypass_results: list[BypassResult],
) -> None:
    """Print a rich terminal report."""
    console = Console()
    console.print()

    # Header
    vendor_str = portal_info.vendor or "Unknown"
    captive_str = "[bold red]YES[/bold red]" if portal_info.is_captive else "[bold green]NO[/bold green]"

    console.print(Panel(
        f"[bold]SSID:[/bold] {portal_info.ssid or 'N/A'}\n"
        f"[bold]Gateway:[/bold] {portal_info.gateway or 'N/A'}\n"
        f"[bold]Captive Portal:[/bold] {captive_str}\n"
        f"[bold]Portal Type:[/bold] {portal_info.portal_type.value}\n"
        f"[bold]Vendor:[/bold] {vendor_str}\n"
        f"[bold]Auth Methods:[/bold] {', '.join(portal_info.auth_methods) or 'N/A'}\n"
        f"[bold]Portal URL:[/bold] {portal_info.portal_url or 'N/A'}",
        title="[bold cyan]nowifi — WiFi Security Audit[/bold cyan]",
        border_style="cyan",
    ))

    # Probe Results
    probe_table = Table(title="Leak Enumeration", border_style="blue")
    probe_table.add_column("Protocol", style="bold")
    probe_table.add_column("Status", justify="center")
    probe_table.add_column("Details")

    probe_table.add_row("DNS (UDP/53)", _bool_icon(probe_results.dns.is_open), probe_results.dns.details)
    probe_table.add_row("ICMP (ping)", _bool_icon(probe_results.icmp.is_open), probe_results.icmp.details)
    probe_table.add_row("IPv6", _bool_icon(probe_results.ipv6.is_open), probe_results.ipv6.details)
    probe_table.add_row("HTTPS (Cloudflare)", _bool_icon(probe_results.cloudflare.is_open), probe_results.cloudflare.details)
    probe_table.add_row("QUIC (UDP/443)", _bool_icon(probe_results.quic.is_open), probe_results.quic.details)
    probe_table.add_row("NTP (UDP/123)", _bool_icon(probe_results.ntp.is_open), probe_results.ntp.details)
    probe_table.add_row("DoH (HTTPS)", _bool_icon(probe_results.doh.is_open), probe_results.doh.details)

    # Whitelisted domains
    for wl in probe_results.whitelists:
        probe_table.add_row(f"  {wl.domain}", _bool_icon(wl.is_open), wl.details)

    # Open ports
    open_ports = [p for p in probe_results.open_ports if p.is_open]
    if open_ports:
        ports_str = ", ".join(f"{p.port}/{p.service}" for p in open_ports)
        probe_table.add_row("Open Ports", f"[green]{len(open_ports)}[/green]", ports_str)
    else:
        probe_table.add_row("Open Ports", "[red]0[/red]", "All scanned ports blocked")

    console.print(probe_table)
    console.print()

    # Bypass Results
    bypass_table = Table(title="Bypass Attempts", border_style="red")
    bypass_table.add_column("Technique", style="bold")
    bypass_table.add_column("Result", justify="center")
    bypass_table.add_column("Severity", justify="center")
    bypass_table.add_column("Impact / Details")

    for r in bypass_results:
        result_str = "[bold green]SUCCESS[/bold green]" if r.success else "[dim]failed[/dim]"
        sev_str = _severity_icon(r.severity) if r.success else "[dim]-[/dim]"
        detail = r.impact if r.success else r.details
        bypass_table.add_row(r.method.value, result_str, sev_str, detail)

    console.print(bypass_table)
    console.print()

    # Findings summary
    successful = [r for r in bypass_results if r.success]
    if successful:
        console.print(Panel(
            "\n".join(
                f"[{_severity_style(r.severity)}]{r.severity.value.upper()}[/{_severity_style(r.severity)}]: "
                f"{r.method.value} — {r.impact}\n"
                f"  [dim]Remediation: {r.remediation}[/dim]"
                for r in successful
            ),
            title=f"[bold red]Findings ({len(successful)} vulnerabilities)[/bold red]",
            border_style="red",
        ))
    else:
        console.print(Panel(
            "[green]No bypass techniques succeeded. Portal appears well-configured.[/green]",
            title="[bold green]No Findings[/bold green]",
            border_style="green",
        ))

    # Active tunnels
    active = [r for r in bypass_results if r.tunnel_handle and r.tunnel_handle.active]
    if active:
        t = active[0]
        console.print(Panel(
            f"[bold green]Tunnel active:[/bold green] {t.method.value}\n"
            f"SOCKS5 proxy: [bold]localhost:{t.tunnel_handle.local_port}[/bold]\n"
            f"Use: [bold]export ALL_PROXY=socks5://127.0.0.1:{t.tunnel_handle.local_port}[/bold]",
            title="[bold green]Active Tunnel[/bold green]",
            border_style="green",
        ))

    console.print()


def generate_markdown_report(
    portal_info: PortalInfo,
    probe_results: ProbeResults,
    bypass_results: list[BypassResult],
) -> str:
    """Generate a markdown pentest report."""
    ts = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M UTC")
    lines = [
        "# Captive Portal Security Assessment Report",
        "",
        f"**Date:** {ts}",
        f"**SSID:** {portal_info.ssid or 'N/A'}",
        f"**Gateway:** {portal_info.gateway or 'N/A'}",
        f"**Portal Vendor:** {portal_info.vendor or 'Unknown'}",
        f"**Portal Type:** {portal_info.portal_type.value}",
        "",
        "## Leak Enumeration",
        "",
        "| Protocol | Status | Details |",
        "|----------|--------|---------|",
        f"| DNS (UDP/53) | {'OPEN' if probe_results.dns.is_open else 'CLOSED'} | {probe_results.dns.details} |",
        f"| ICMP | {'OPEN' if probe_results.icmp.is_open else 'CLOSED'} | {probe_results.icmp.details} |",
        f"| IPv6 | {'OPEN' if probe_results.ipv6.is_open else 'CLOSED'} | {probe_results.ipv6.details} |",
        f"| HTTPS (CF) | {'OPEN' if probe_results.cloudflare.is_open else 'CLOSED'} | {probe_results.cloudflare.details} |",
        f"| QUIC (UDP/443) | {'OPEN' if probe_results.quic.is_open else 'CLOSED'} | {probe_results.quic.details} |",
        f"| NTP (UDP/123) | {'OPEN' if probe_results.ntp.is_open else 'CLOSED'} | {probe_results.ntp.details} |",
        f"| DoH (HTTPS) | {'OPEN' if probe_results.doh.is_open else 'CLOSED'} | {probe_results.doh.details} |",
    ]

    for wl in probe_results.whitelists:
        lines.append(f"| {wl.domain} | {'OPEN' if wl.is_open else 'CLOSED'} | {wl.details} |")

    lines += [
        "",
        "## Bypass Results",
        "",
        "| Technique | Result | Severity | Impact |",
        "|-----------|--------|----------|--------|",
    ]

    for r in bypass_results:
        result = "SUCCESS" if r.success else "failed"
        sev = r.severity.value.upper() if r.success else "-"
        detail = r.impact if r.success else r.details
        lines.append(f"| {r.method.value} | {result} | {sev} | {detail} |")

    successful = [r for r in bypass_results if r.success]
    if successful:
        lines += ["", "## Findings & Remediation", ""]
        for i, r in enumerate(successful, 1):
            lines += [
                f"### {i}. {r.method.value} ({r.severity.value.upper()})",
                "",
                f"**Impact:** {r.impact}",
                "",
                f"**Details:** {r.details}",
                "",
                f"**Remediation:** {r.remediation}",
                "",
            ]

    lines += [
        "---",
        "*Generated by nowifi v0.1.0*",
    ]

    return "\n".join(lines)


def generate_json_report(
    portal_info: PortalInfo,
    probe_results: ProbeResults,
    bypass_results: list[BypassResult],
) -> str:
    """Generate NDJSON report."""
    report = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "portal": {
            "is_captive": portal_info.is_captive,
            "type": portal_info.portal_type.value,
            "vendor": portal_info.vendor,
            "ssid": portal_info.ssid,
            "gateway": portal_info.gateway,
            "url": portal_info.portal_url,
            "auth_methods": portal_info.auth_methods,
        },
        "probes": {
            "dns": {"open": probe_results.dns.is_open, "details": probe_results.dns.details},
            "icmp": {"open": probe_results.icmp.is_open, "details": probe_results.icmp.details},
            "ipv6": {"open": probe_results.ipv6.is_open, "details": probe_results.ipv6.details},
            "cloudflare": {"open": probe_results.cloudflare.is_open, "details": probe_results.cloudflare.details},
            "quic": {"open": probe_results.quic.is_open, "details": probe_results.quic.details},
            "ntp": {"open": probe_results.ntp.is_open, "details": probe_results.ntp.details},
            "doh": {"open": probe_results.doh.is_open, "details": probe_results.doh.details},
            "whitelists": [{"domain": w.domain, "open": w.is_open} for w in probe_results.whitelists],
            "open_ports": [{"port": p.port, "service": p.service} for p in probe_results.open_ports if p.is_open],
        },
        "bypasses": [
            {
                "method": r.method.value,
                "success": r.success,
                "severity": r.severity.value,
                "impact": r.impact,
                "details": r.details,
                "remediation": r.remediation,
            }
            for r in bypass_results
        ],
    }
    return json.dumps(report, indent=2)

File: go/internal/report/report.go
// Copyright (C) 2026 Mikko Parkkola. All rights reserved.
// Licensed under AGPL-3.0. See LICENSE file.

// Package report generates terminal, markdown, and JSON output for
// captive portal security audit results.
//
// Terminal output uses fatih/color for colored text. The format closely
// mirrors the Python Rich-based output: header panel, leak enumeration
// table, bypass results table, findings summary, and active tunnel info.
package report

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/MikkoParkkola/nowifi/internal/bypass"
	"golang.org/x/term"
)

// ---------------------------------------------------------------------------
// Types consumed by the report generators. These mirror the detect and
// probe packages' types so report can compile independently. The caller
// maps from the concrete types to these.
// ---------------------------------------------------------------------------

// PortalInfo summarizes the detected captive portal.
type PortalInfo struct {
	IsCaptive   bool
	PortalType  string // e.g. "http_redirect", "dns_hijack", "firewall_block"
	Vendor      string
	SSID        string
	Gateway     string
	PortalURL   string
	AuthMethods []string
}

// ProbeResult holds the status of a single protocol probe.
type ProbeResult struct {
	IsOpen  bool
	Details string
}

// WhitelistResult holds a probed whitelisted domain.
type WhitelistResult struct {
	Domain  string
	IsOpen  bool
	Details string
}

// PortResult holds a probed TCP port.
type PortResult struct {
	Port    int
	Service string
	IsOpen  bool
}

// ProbeResults is the full set of probe outcomes for reporting.
type ProbeResults struct {
	DNS        ProbeResult
	ICMP       ProbeResult
	IPv6       ProbeResult
	Cloudflare ProbeResult
	QUIC       ProbeResult
	NTP        ProbeResult
	DoH        ProbeResult

	Whitelists []WhitelistResult
	OpenPorts  []PortResult
}

// ---------------------------------------------------------------------------
// Color helpers (ANSI escape codes)
// ---------------------------------------------------------------------------

const (
	reset     = "\033[0m"
	bold      = "\033[1m"
	dim       = "\033[2m"
	red       = "\033[31m"
	green     = "\033[32m"
	yellow    = "\033[33m"
	cyan      = "\033[36m"
	boldRed   = "\033[1;31m"
	boldGreen = "\033[1;32m"
	boldCyan  = "\033[1;36m"
)

var colorEnabled = os.Getenv("NO_COLOR") == "" && term.IsTerminal(int(os.Stdout.Fd()))

func renderTerminalOutput(s string) string {
	if colorEnabled {
		return s
	}
	return stripANSI(s)
}

func severityColor(sev string) string {
	switch sev {
	case "critical":
		return boldRed
	case "high":
		return red
	case "medium":
		return yellow
	case "low":
		return cyan
	case "info":
		return dim
	default:
		return dim
	}
}

func severityTag(sev string) string {
	switch sev {
	case "critical":
		return boldRed + "CRIT" + reset
	case "high":
		return red + "HIGH" + reset
	case "medium":
		return yellow + "MED" + reset
	case "low":
		return cyan + "LOW" + reset
	case "info":
		return dim + "INFO" + reset
	default:
		return dim + "?" + reset
	}
}

func boolIcon(val bool) string {
	if val {
		return green + "OPEN" + reset
	}
	return red + "CLOSED" + reset
}

func successIcon(val bool) string {
	if val {
		return boldGreen + "SUCCESS" + reset
	}
	return dim + "failed" + reset
}

// ---------------------------------------------------------------------------
// Box-drawing helpers for terminal panels
// ---------------------------------------------------------------------------

func drawBox(title string, borderColor string, content string) string {
	lines := strings.Split(content, "\n")
	maxWidth := len(title) + 4
	for _, l := range lines {
		// Approximate visible length (strip ANSI codes).
		vis := stripANSI(l)
		if len(vis) > maxWidth {
			maxWidth = len(vis)
		}
	}
	if maxWidth > 100 {
		maxWidth = 100
	}

	var sb strings.Builder
	topLeft := borderColor + "+" + strings.Repeat("-", maxWidth+2) + "+" + reset
	sb.WriteString(topLeft + "\n")

	// Title line.
	titlePad := maxWidth - len(stripANSI(title))
	if titlePad < 0 {
		titlePad = 0
	}
	sb.WriteString(borderColor + "| " + reset + title + strings.Repeat(" ", titlePad) + borderColor + " |" + reset + "\n")
	sb.WriteString(borderColor + "+" + strings.Repeat("-", maxWidth+2) + "+" + reset + "\n")

	// Content lines.
	for _, l := range lines {
		vis := stripANSI(l)
		pad := maxWidth - len(vis)
		if pad < 0 {
			pad = 0
		}
		sb.WriteString(borderColor + "| " + reset + l + strings.Repeat(" ", pad) + borderColor + " |" + reset + "\n")
	}

	sb.WriteString(borderColor + "+" + strings.Repeat("-", maxWidth+2) + "+" + reset + "\n")
	return renderTerminalOutput(sb.String())
}

// stripANSI removes ANSI escape sequences for length calculation.
func stripANSI(s string) string {
	var result strings.Builder
	inEscape := false
	for _, c := range s {
		if c == '\033' {
			inEscape = true
			continue
		}
		if inEscape {
			if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
				inEscape = false
			}
			continue
		}
		result.WriteRune(c)
	}
	return result.String()
}

// ---------------------------------------------------------------------------
// Table rendering
// ---------------------------------------------------------------------------

// simpleTable renders an ASCII table with headers and rows.
func simpleTable(title string, borderColor string, headers []string, rows [][]string) string {
	// Compute column widths.
	widths := make([]int, len(headers))
	for i, h := range headers {
		widths[i] = len(stripANSI(h))
	}
	for _, row := range rows {
		for i, cell := range row {
			if i < len(widths) {
				vis := len(stripANSI(cell))
				if vis > widths[i] {
					widths[i] = vis
				}
			}
		}
	}

	sep := borderColor + "+"
	for _, w := range widths {
		sep += strings.Repeat("-", w+2) + "+"
	}
	sep += reset

	var sb strings.Builder

	// Title.
	if title != "" {
		sb.WriteString("\n" + bold + title + reset + "\n")
	}

	sb.WriteString(sep + "\n")

	// Header row.
	sb.WriteString(borderColor + "|" + reset)
	for i, h := range headers {
		pad := widths[i] - len(stripANSI(h))
		sb.WriteString(" " + bold + h + reset + strings.Repeat(" ", pad) + " " + borderColor + "|" + reset)
	}
	sb.WriteString("\n" + sep + "\n")

	// Data rows.
	for _, row := range rows {
		sb.WriteString(borderColor + "|" + reset)
		for i := 0; i < len(headers); i++ {
			cell := ""
			if i < len(row) {
				cell = row[i]
			}
			vis := len(stripANSI(cell))
			pad := widths[i] - vis
			if pad < 0 {
				pad = 0
			}
			sb.WriteString(" " + cell + strings.Repeat(" ", pad) + " " + borderColor + "|" + reset)
		}
		sb.WriteString("\n")
	}

	sb.WriteString(sep + "\n")
	return renderTerminalOutput(sb.String())
}

// ---------------------------------------------------------------------------
// Terminal report
// ---------------------------------------------------------------------------

// PrintTerminal prints a colored terminal report to stdout.
func PrintTerminal(portal PortalInfo, probes ProbeResults, bypasses []bypass.Result) {
	// Header panel.
	vendorStr := portal.Vendor
	if vendorStr == "" {
		vendorStr = "Unknown"
	}
	captiveStr := boldRed + "YES" + reset
	if !portal.IsCaptive {
		captiveStr = boldGreen + "NO" + reset
	}
	authStr := "N/A"
	if len(portal.AuthMethods) > 0 {
		authStr = strings.Join(portal.AuthMethods, ", ")
	}
	portalURLStr := portal.PortalURL
	if portalURLStr == "" {
		portalURLStr = "N/A"
	}

	header := fmt.Sprintf(
		"%sSSID:%s %s\n%sGateway:%s %s\n%sCaptive Portal:%s %s\n%sPortal Type:%s %s\n%sVendor:%s %s\n%sAuth Methods:%s %s\n%sPortal URL:%s %s",
		bold, reset, nvl(portal.SSID, "N/A"),
		bold, reset, nvl(portal.Gateway, "N/A"),
		bold, reset, captiveStr,
		bold, reset, portal.PortalType,
		bold, reset, vendorStr,
		bold, reset, authStr,
		bold, reset, portalURLStr,
	)
	fmt.Println()
	fmt.Print(drawBox(boldCyan+"nowifi -- WiFi Security Audit"+reset, cyan, header))

	// Probe results table.
	probeRows := [][]string{
		{"DNS (UDP/53)", boolIcon(probes.DNS.IsOpen), probes.DNS.Details},
		{"ICMP (ping)", boolIcon(probes.ICMP.IsOpen), probes.ICMP.Details},
		{"IPv6", boolIcon(probes.IPv6.IsOpen), probes.IPv6.Details},
		{"HTTPS (Cloudflare)", boolIcon(probes.Cloudflare.IsOpen), probes.Cloudflare.Details},
		{"QUIC (UDP/443)", boolIcon(probes.QUIC.IsOpen), probes.QUIC.Details},
		{"NTP (UDP/123)", boolIcon(probes.NTP.IsOpen), probes.NTP.Details},
		{"DoH (HTTPS)", boolIcon(probes.DoH.IsOpen), probes.DoH.Details},
	}

	for _, wl := range probes.Whitelists {
		probeRows = append(probeRows, []string{"  " + wl.Domain, boolIcon(wl.IsOpen), wl.Details})
	}

	var openPorts []PortResult
	for _, p := range probes.OpenPorts {
		if p.IsOpen {
			openPorts = append(openPorts, p)
		}
	}
	if len(openPorts) > 0 {
		var portStrs []string
		for _, p := range openPorts {
			portStrs = append(portStrs, fmt.Sprintf("%d/%s", p.Port, p.Service))
		}
		probeRows = append(probeRows, []string{
			"Open Ports",
			green + fmt.Sprintf("%d", len(openPorts)) + reset,
			strings.Join(portStrs, ", "),
		})
	} else {
		probeRows = append(probeRows, []string{
			"Open Ports",
			red + "0" + reset,
			"All scanned ports blocked",
		})
	}

	fmt.Print(simpleTable("Leak Enumeration", "\033[34m", []string{"Protocol", "Status", "Details"}, probeRows))

	// Bypass results table.
	var bypassRows [][]string
	for _, r := range bypasses {
		resultStr := successIcon(r.Success)
		sevStr := dim + "-" + reset
		if r.Success {
			sevStr = severityTag(r.Severity)
		}
		detail := r.Details
		if r.Success {
			detail = r.Impact
		}
		bypassRows = append(bypassRows, []string{string(r.Method), resultStr, sevStr, detail})
	}

	fmt.Print(simpleTable("Bypass Attempts", red, []string{"Technique", "Result", "Severity", "Impact / Details"}, bypassRows))
	fmt.Println()

	// Findings summary.
	var successful []bypass.Result
	for _, r := range bypasses {
		if r.Success {
			successful = append(successful, r)
		}
	}

	if len(successful) > 0 {
		var findings strings.Builder
		for _, r := range successful {
			sc := severityColor(r.Severity)
			findings.WriteString(fmt.Sprintf(
				"%s%s%s: %s -- %s\n  %sRemediation: %s%s\n",
				sc, strings.ToUpper(r.Severity), reset,
				string(r.Method), r.Impact,
				dim, r.Remediation, reset,
			))
		}
		fmt.Print(drawBox(
			boldRed+fmt.Sprintf("Findings (%d vulnerabilities)", len(successful))+reset,
			red, findings.String(),
		))
	} else {
		fmt.Print(drawBox(
			boldGreen+"No Findings"+reset,
			green,
			green+"No bypass techniques succeeded. Portal appears well-configured."+reset,
		))
	}

	// Active tunnel info.
	for _, r := range bypasses {
		if r.Tunnel != nil && r.Tunnel.Active {
			tunnelInfo := fmt.Sprintf(
				"%sTunnel active:%s %s\nSOCKS5 proxy: %slocalhost:%d%s\nUse: %sexport ALL_PROXY=socks5://127.0.0.1:%d%s",
				boldGreen, reset, string(r.Method),
				bold, r.Tunnel.LocalPort, reset,
				bold, r.Tunnel.LocalPort, reset,
			)
			fmt.Print(drawBox(boldGreen+"Active Tunnel"+reset, green, tunnelInfo))
			break
		}
	}

	fmt.Println()
}

// ---------------------------------------------------------------------------
// Markdown report
// ---------------------------------------------------------------------------

// GenerateMarkdown produces a markdown pentest report as a string.
func GenerateMarkdown(portal PortalInfo, probes ProbeResults, bypasses []bypass.Result) string {
	ts := time.Now().UTC().Format("2006-01-02 15:04 UTC")

	var b strings.Builder
	b.WriteString("# Captive Portal Security Assessment Report\n\n")
	b.WriteString(fmt.Sprintf("**Date:** %s\n", ts))
	b.WriteString(fmt.Sprintf("**SSID:** %s\n", nvl(portal.SSID, "N/A")))
	b.WriteString(fmt.Sprintf("**Gateway:** %s\n", nvl(portal.Gateway, "N/A")))
	b.WriteString(fmt.Sprintf("**Portal Vendor:** %s\n", nvl(portal.Vendor, "Unknown")))
	b.WriteString(fmt.Sprintf("**Portal Type:** %s\n", portal.PortalType))
	b.WriteString("\n## Leak Enumeration\n\n")
	b.WriteString("| Protocol | Status | Details |\n")
	b.WriteString("|----------|--------|--------|\n")

	writeProbeRow := func(name string, pr ProbeResult) {
		status := "CLOSED"
		if pr.IsOpen {
			status = "OPEN"
		}
		b.WriteString(fmt.Sprintf("| %s | %s | %s |\n", name, status, pr.Details))
	}

	writeProbeRow("DNS (UDP/53)", probes.DNS)
	writeProbeRow("ICMP", probes.ICMP)
	writeProbeRow("IPv6", probes.IPv6)
	writeProbeRow("HTTPS (CF)", probes.Cloudflare)
	writeProbeRow("QUIC (UDP/443)", probes.QUIC)
	writeProbeRow("NTP (UDP/123)", probes.NTP)
	writeProbeRow("DoH (HTTPS)", probes.DoH)

	for _, wl := range probes.Whitelists {
		status := "CLOSED"
		if wl.IsOpen {
			status = "OPEN"
		}
		b.WriteString(fmt.Sprintf("| %s | %s | %s |\n", wl.Domain, status, wl.Details))
	}

	b.WriteString("\n## Bypass Results\n\n")
	b.WriteString("| Technique | Result | Severity | Impact |\n")
	b.WriteString("|-----------|--------|----------|--------|\n")

	for _, r := range bypasses {
		result := "failed"
		if r.Success {
			result = "SUCCESS"
		}
		sev := "-"
		if r.Success {
			sev = strings.ToUpper(r.Severity)
		}
		detail := r.Details
		if r.Success {
			detail = r.Impact
		}
		b.WriteString(fmt.Sprintf("| %s | %s | %s | %s |\n", string(r.Method), result, sev, detail))
	}

	var successful []bypass.Result
	for _, r := range bypasses {
		if r.Success {
			successful = append(successful, r)
		}
	}

	if len(successful) > 0 {
		b.WriteString("\n## Findings & Remediation\n\n")
		for i, r := range successful {
			b.WriteString(fmt.Sprintf("### %d. %s (%s)\n\n", i+1, string(r.Method), strings.ToUpper(r.Severity)))
			b.WriteString(fmt.Sprintf("**Impact:** %s\n\n", r.Impact))
			b.WriteString(fmt.Sprintf("**Details:** %s\n\n", r.Details))
			b.WriteString(fmt.Sprintf("**Remediation:** %s\n\n", r.Remediation))
		}
	}

	b.WriteString("---\n*Generated by nowifi v0.1.0*\n")
	return b.String()
}

// ---------------------------------------------------------------------------
// JSON report
// ---------------------------------------------------------------------------

// jsonReport is the serialization structure for JSON output.
type jsonReport struct {
	Timestamp string       `json:"timestamp"`
	Portal    jsonPortal   `json:"portal"`
	Probes    jsonProbes   `json:"probes"`
	Bypasses  []jsonBypass `json:"bypasses"`
}

type jsonPortal struct {
	IsCaptive   bool     `json:"is_captive"`
	Type        string   `json:"type"`
	Vendor      string   `json:"vendor"`
	SSID        string   `json:"ssid"`
	Gateway     string   `json:"gateway"`
	URL         string   `json:"url"`
	AuthMethods []string `json:"auth_methods"`
}

type jsonProbeEntry struct {
	Open    bool   `json:"open"`
	Details string `json:"details"`
}

type jsonWhitelist struct {
	Domain string `json:"domain"`
	Open   bool   `json:"open"`
}

type jsonOpenPort struct {
	Port    int    `json:"port"`
	Service string `json:"service"`
}

type jsonProbes struct {
	DNS        jsonProbeEntry  `json:"dns"`
	ICMP       jsonProbeEntry  `json:"icmp"`
	IPv6       jsonProbeEntry  `json:"ipv6"`
	Cloudflare jsonProbeEntry  `json:"cloudflare"`
	QUIC       jsonProbeEntry  `json:"quic"`
	NTP        jsonProbeEntry  `json:"ntp"`
	DoH        jsonProbeEntry  `json:"doh"`
	Whitelists []jsonWhitelist `json:"whitelists"`
	OpenPorts  []jsonOpenPort  `json:"open_ports"`
}

type jsonBypass struct {
	Method      string `json:"method"`
	Success     bool   `json:"success"`
	Severity    string `json:"severity"`
	Impact      string `json:"impact"`
	Details     string `json:"details"`
	Remediation string `json:"remediation"`
}

// GenerateJSON produces a JSON report as a formatted string.
func GenerateJSON(portal PortalInfo, probes ProbeResults, bypasses []bypass.Result) string {
	report := jsonReport{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Portal: jsonPortal{
			IsCaptive:   portal.IsCaptive,
			Type:        portal.PortalType,
			Vendor:      portal.Vendor,
			SSID:        portal.SSID,
			Gateway:     portal.Gateway,
			URL:         portal.PortalURL,
			AuthMethods: portal.AuthMethods,
		},
		Probes: jsonProbes{
			DNS:        jsonProbeEntry{Open: probes.DNS.IsOpen, Details: probes.DNS.Details},
			ICMP:       jsonProbeEntry{Open: probes.ICMP.IsOpen, Details: probes.ICMP.Details},
			IPv6:       jsonProbeEntry{Open: probes.IPv6.IsOpen, Details: probes.IPv6.Details},
			Cloudflare: jsonProbeEntry{Open: probes.Cloudflare.IsOpen, Details: probes.Cloudflare.Details},
			QUIC:       jsonProbeEntry{Open: probes.QUIC.IsOpen, Details: probes.QUIC.Details},
			NTP:        jsonProbeEntry{Open: probes.NTP.IsOpen, Details: probes.NTP.Details},
			DoH:        jsonProbeEntry{Open: probes.DoH.IsOpen, Details: probes.DoH.Details},
		},
	}

	for _, wl := range probes.Whitelists {
		report.Probes.Whitelists = append(report.Probes.Whitelists, jsonWhitelist{
			Domain: wl.Domain,
			Open:   wl.IsOpen,
		})
	}

	for _, p := range probes.OpenPorts {
		if p.IsOpen {
			report.Probes.OpenPorts = append(report.Probes.OpenPorts, jsonOpenPort{
				Port:    p.Port,
				Service: p.Service,
			})
		}
	}

	for _, r := range bypasses {
		report.Bypasses = append(report.Bypasses, jsonBypass{
			Method:      string(r.Method),
			Success:     r.Success,
			Severity:    r.Severity,
			Impact:      r.Impact,
			Details:     r.Details,
			Remediation: r.Remediation,
		})
	}

	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return fmt.Sprintf(`{"error": "%s"}`, err.Error())
	}
	return string(data)
}

// nvl returns val if non-empty, otherwise fallback.
func nvl(val, fallback string) string {
	if val == "" {
		return fallback
	}
	return val
}
